src/main/java/com/mraof/minestuck/data/worldgen/MSPlacedFeatureProvider.java

```diff
@@ -53,17 +53,18 @@ public static void register(BootstapContext<PlacedFeature> context)
 		context.register(UNCOMMON_BROKEN_SWORD, placed(features, MSCFeatures.BROKEN_SWORD,
 				worldGenModifiers(RarityFilter.onAverageOnceEvery(50))));
 		context.register(BUCKET, placed(features, MSCFeatures.BUCKET,
-				worldGenModifiers(RarityFilter.onAverageOnceEvery(16), PlacementUtils.HEIGHTMAP_WORLD_SURFACE)));
+				worldGenModifiers(RarityFilter.onAverageOnceEvery(16), PlacementUtils.HEIGHTMAP_WORLD_SURFACE,
+						BlockPredicateFilter.forPredicate(BlockPredicate.noFluid(Direction.DOWN.getNormal())), RandomOffsetPlacement.vertical(UniformInt.of(-2, 0)))));
 		context.register(CAKE_PEDESTAL, placed(features, MSCFeatures.CAKE_PEDESTAL,
-				worldGenModifiers(RarityFilter.onAverageOnceEvery(100))));
+				worldGenModifiers(RarityFilter.onAverageOnceEvery(100), PlacementUtils.HEIGHTMAP_WORLD_SURFACE)));
 		context.register(SMALL_LIBRARY, placed(features, MSCFeatures.SMALL_LIBRARY,
 				worldGenModifiers(RarityFilter.onAverageOnceEvery(64))));
 		context.register(TOWER, placed(features, MSCFeatures.TOWER,
 				worldGenModifiers(RarityFilter.onAverageOnceEvery(20), PlacementUtils.HEIGHTMAP_WORLD_SURFACE)));
 		context.register(PARCEL_PYXIS, placed(features, MSCFeatures.PARCEL_PYXIS,
-				worldGenModifiers(RarityFilter.onAverageOnceEvery(60), PlacementUtils.HEIGHTMAP)));
+				worldGenModifiers(RarityFilter.onAverageOnceEvery(60), PlacementUtils.HEIGHTMAP, BlockPredicateFilter.forPredicate(BlockPredicate.noFluid()))));
 		context.register(FROG_RUINS, placed(features, MSCFeatures.FROG_RUINS,
-				worldGenModifiers(RarityFilter.onAverageOnceEvery(60), PlacementUtils.HEIGHTMAP_OCEAN_FLOOR)));
+				worldGenModifiers(RarityFilter.onAverageOnceEvery(60), PlacementUtils.HEIGHTMAP_OCEAN_FLOOR, RandomOffsetPlacement.vertical(ConstantInt.of(-3)))));
```

Comment:
Oh, right. From what I saw, the original one didn't actually have a range due to a programmer error (inclusive vs. exclusive range, off by one error). However, since the difference is aesthetic more than anything else, I chose to follow the actual behaviour rather than the intent.

If you want, I could go ahead with the intent instead. I think there was one other feature that had a range like that too.